[RFC] Remove "j" flag from 'cpoptions'

glts opened the pull request:
```
                                                            *cpo-j*
            j       When joining lines, only add two spaces after a '.',
                    not after '!' or '?'.  Also see 'joinspaces'.
```

I don't see the point of special-casing '!' and '?' when double-spacing. I think this option can be safely removed.

Since test29 used this flag, I fixed it and took the opportunity to convert it directly to Lua.

glts committed:
'cpoptions': Remove "j" flag

glts committed:
tests: Convert legacy test29

Pyrohh commented:
Could you please rebase? I'll definitely take a look after, but what do you think about leaving tests for separate PRs in the future?

I rather like how small the `cpo` PRs have been, and think that adding legacy test conversions to them slows down their mean time to merge (MTTM, I might use that... :smile:).

glts commented:
Sure, done.

Yeah, sorry about the test. The problem was that the test would fail given that it tested `cpoptions+=j` specifically. I would have had to touch the `test29.{in,ok}` files anyway.

I can still split the changes and move the test conversion to a separate pull request. Just let me know :smile:.

Pyrohh commented:
> Yeah, sorry about the test. The problem was that the test would fail given that it tested cpoptions+=j specifically. I would have had to touch the test29.{in,ok} files anyway.

Ah okay, well I suppose that's fine.

Pyrohh commented:
Welp, looks like you need to rebase again.

I hate `cpoptions`.

glts commented:
Rebased.

Pyrohh commented on runtime/doc/options.txt:
```diff
@@ -1779,9 +1779,6 @@ A jump table for the options with a short description can be found at |Q_op|.
 		I	When moving the cursor up or down just after inserting
 			indent for 'autoindent', do not delete the indent.
-								*cpo-j*
-		j	When joining lines, only add two spaces after a '.',
```
See `:h joinspaces`, there's still a reference to this.

glts replied:
Good catch, I'll just remove the line

        When 'cpoptions' includes the 'j' flag, only do this after a '.'.

Pyrohh commented on src/nvim/ops.c:
```diff
@@ -3460,12 +3460,10 @@ int do_join(long count,
-                       || (vim_strchr(p_cpo, CPO_JOINSP) == NULL
-                           && (endcurr1 == '?' || endcurr1 == '!'))))
+        // Extra space when 'joinspaces' set and line ends in '.', '?', or '!'.
+        if (p_js && (endcurr1 == '.' || endcurr1 == '?' || endcurr1 == '!')) {
```
perhaps a line break after the `&&` like above (`edit.c`)

glts replied:
The conditional expression is quite clear and the line fits in 80 columns. Is there a reason you'd prefer it split?

Pyrohh replied:
As I said, to match what's done in `edit.c`. Anyways, I think you're right that it's clear enough :+1:

Pyrohh commented:
I skimmed the tests and tried them out (modifying pieces of code that would break the tests), and they seem fine to me. Still, I don't know much regarding the tests so take my approval with a grain of salt. Besides the tests and the two nits, LGTM.

glts commented:
Rebased with only the one-line removal in `runtime/doc/options.txt`.

Pyrohh commented:
There's a reference to this in `change.txt`, line 150.

Anyways I took another look and the tests work fine (as does everything else). Besides the small nits I mentioned, and a rebase to clear the merge conflict (it's in `CPO_{VI,ALL}` again :sweat:), this LGTM.

glts committed:
fixup! tests

glts commented:
First rebased with only the conflict in `option_defs.h`.

Then I've added a review commit for test29. I've changed the wording of some of the comments and `describe` titles to make more sense considering our default settings (`joinspaces` and `formatoptions+=j`).

glts commented:
Thanks for the review, @Pyrohh.

glts committed:
fixup! 'cpoptions'

glts commented:
Had to rebase due to a conflict in the legacy test Makefile.

I've added another commit to remove an obsolete remark in `change.txt` (as noted by @Pyrohh, I had overlooked this).

justinmk commented:
Merged. Great work!

glts commented:
Thanks @justinmk.
